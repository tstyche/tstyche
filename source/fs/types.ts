export interface FileSystemEntries {
  directories: Array<string>;
  files: Array<string>;
}

export interface FileSystemEntryMeta {
  isDirectory: () => boolean;
  isFile: () => boolean;
  isSymbolicLink: () => boolean;
}

export type FileTree = Record<string, string>;
