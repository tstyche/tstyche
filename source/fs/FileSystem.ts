import { readFileSync, readdirSync, statSync } from "node:fs";
import type { MemoryFiles } from "./MemoryFiles.js";
import type { FileSystemEntries, FileSystemEntryMeta } from "./types.js";

export class FileSystem {
  static #memoryFiles: MemoryFiles | undefined;

  static addMemoryFiles(memoryFiles: MemoryFiles) {
    FileSystem.#memoryFiles = memoryFiles;
  }

  static directoryExists(path: string): boolean {
    return FileSystem.#memoryFiles?.hasDirectory(path) || !!statSync(path, { throwIfNoEntry: false })?.isDirectory();
  }

  static getAccessibleFileSystemEntries(path: string): FileSystemEntries {
    const directories: Array<string> = [];
    const files: Array<string> = [];

    try {
      const entries = readdirSync(path, { withFileTypes: true });

      for (const entry of entries) {
        let entryMeta: FileSystemEntryMeta | undefined = entry;

        if (entry.isSymbolicLink()) {
          entryMeta = statSync([path, entry.name].join("/"), { throwIfNoEntry: false });
        }

        if (entryMeta?.isDirectory() === true) {
          directories.push(entry.name);
        } else if (entryMeta?.isFile() === true) {
          files.push(entry.name);
        }
      }
    } catch {
      // continue regardless of error
    }

    if (!FileSystem.#memoryFiles) {
      return { directories, files };
    }

    const memoryEntries = FileSystem.#memoryFiles.getEntries(path);

    return {
      directories: [...new Set(...directories, ...memoryEntries.directories)].sort(),
      files: [...new Set(...files, ...memoryEntries.files)].sort(),
    };
  }

  static fileExists(path: string): boolean {
    return FileSystem.#memoryFiles?.hasFile(path) || !!statSync(path, { throwIfNoEntry: false })?.isFile();
  }

  static readFile(path: string): string | undefined {
    let contents = FileSystem.#memoryFiles?.getFile(path);

    if (contents != null) {
      return contents;
    }

    try {
      contents = readFileSync(path, { encoding: "utf8" });
    } catch {
      // continue regardless of error
    }

    return contents;
  }

  static removeMemoryFiles() {
    FileSystem.#memoryFiles = undefined;
  }
}
