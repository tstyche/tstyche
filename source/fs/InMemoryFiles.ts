import { Path } from "#path";
import type { FileSystemEntries, FileTree } from "./types.js";

export class InMemoryFiles {
  #directories = new Map<string, { directories: Set<string>; files: Set<string> }>();
  #files = new Map<string, string>();
  #rootPath: string;

  constructor(rootPath: string) {
    this.#rootPath = Path.resolve(rootPath);
  }

  add(fileTree: FileTree): InMemoryFiles {
    for (const fileKey in fileTree) {
      const filePath = Path.relative(this.#rootPath, fileKey);

      if (!filePath.startsWith("./")) {
        // TODO log a warning first, perhaps via 'fs:add' or 'files:add'?
        continue;
      }

      const pathSegments = filePath.split("/").filter((pathSegment) => pathSegment !== ".");

      let currentPath = this.#rootPath;
      let pathSegmentIndex = 0;

      for (const pathSegment of pathSegments) {
        let entries = this.#directories.get(currentPath);

        if (!entries) {
          entries = { directories: new Set(), files: new Set() };
          this.#directories.set(currentPath, entries);
        }

        currentPath = Path.join(currentPath, pathSegment);
        ++pathSegmentIndex;

        if (pathSegmentIndex === pathSegments.length) {
          entries.files.add(pathSegment);

          // biome-ignore lint/style/noNonNullAssertion: this is fine
          this.#files.set(currentPath, fileTree[fileKey]!);
        } else {
          entries.directories.add(pathSegment);
        }
      }
    }

    return this;
  }

  getEntries(path: string): FileSystemEntries {
    const entries = this.#directories.get(path);

    if (!entries) {
      return { directories: [], files: [] };
    }

    return { directories: [...entries.directories], files: [...entries.files] };
  }

  getFile(path: string): string | undefined {
    return this.#files.get(path);
  }

  hasDirectory(path: string): boolean {
    return this.#directories.has(path);
  }

  hasFile(path: string): boolean {
    return this.#files.has(path);
  }
}
