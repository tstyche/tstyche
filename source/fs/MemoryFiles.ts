import path from "node:path";
import { Path } from "#path";
import type { FileSystemEntries } from "./types.js";

export class MemoryFiles {
  #directories = new Map<string, { directories: Set<string>; files: Set<string> }>();
  #files = new Map<string, string>();
  #rootPath: string;

  constructor(rootPath: string) {
    this.#rootPath = rootPath;
  }

  add(files: Record<string, string>) {
    for (const filePath in files) {
      if (path.isAbsolute(filePath)) {
        Path.relative(this.#rootPath, filePath);
      }

      if (!filePath.startsWith(".")) {
        // TODO log a warning fist, perhaps via 'fs:info' or general 'warnings' event?
        continue;
      }

      const pathSegments = filePath.split("/");
      let pathSegmentIndex = 0;
      let currentPath = this.#rootPath;

      for (const pathSegment of pathSegments) {
        ++pathSegmentIndex;

        let entries = this.#directories.get(currentPath);

        if (!entries) {
          entries = { directories: new Set(), files: new Set() };
          this.#directories.set(currentPath, entries);
        }

        if (pathSegmentIndex === pathSegments.length) {
          entries.files.add(currentPath);

          // biome-ignore lint/style/noNonNullAssertion: this is fine
          this.#files.set(filePath, files[filePath]!);
        } else {
          entries.directories.add(currentPath);
        }

        currentPath = Path.join(currentPath, pathSegment);
      }
    }
  }

  getEntries(path: string): FileSystemEntries {
    const entries = this.#directories.get(path);

    if (!entries) {
      return { directories: [], files: [] };
    }

    return { directories: [...entries.directories], files: [...entries.files] };
  }

  getFile(path: string): string | undefined {
    return this.#files.get(path);
  }

  hasDirectory(path: string): boolean {
    return this.#directories.has(path);
  }

  hasFile(path: string): boolean {
    return this.#files.has(path);
  }
}
