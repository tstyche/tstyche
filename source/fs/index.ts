export { FileSystem } from "./FileSystem.js";
export { InMemoryFiles } from "./InMemoryFiles.js";
export type { FileTree } from "./types.js";
