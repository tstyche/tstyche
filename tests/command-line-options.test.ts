import { existsSync } from "node:fs";
import { afterEach, describe, expect, test } from "@jest/globals";
import { clearFixture, writeFixture } from "./__utils__/fixtureFactory.js";
import { getFixtureUrl } from "./__utils__/getFixtureUrl.js";
import { spawnTyche } from "./__utils__/spawnTyche.js";

const isStringTestText = `import { expect, test } from "tstyche";
test("is string?", () => {
  expect<string>().type.toBeString();
});
`;

const tsconfig = {
  extends: "../tsconfig.json",
  include: ["./"],
};

const fixture = "command-line-options";

afterEach(async () => {
  await clearFixture(fixture);
});

describe("command line options", () => {
  describe("'--install' option", () => {
    test("with 'target' configuration option", async () => {
      const config = { target: ["4.8", "5.0"] };

      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: isStringTestText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
        ["tstyche.config.json"]: JSON.stringify(config, null, 2),
      });

      const { status, stderr, stdout } = spawnTyche(fixture, ["--install"], { ["TSTYCHE_STORE_PATH"]: "./.store" });

      expect(stdout).toMatchInlineSnapshot(`
        "adds TypeScript 4.8.4 to <<cwd>>/tests/__fixtures__/command-line-options/.store/4.8.4
        adds TypeScript 5.0.4 to <<cwd>>/tests/__fixtures__/command-line-options/.store/5.0.4
        "
      `);
      expect(stderr).toBe("");

      expect(status).toBe(0);
    });

    test("with '--target' command line options", async () => {
      const config = { target: ["5.0", "latest"] };

      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: isStringTestText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
        ["tstyche.config.json"]: JSON.stringify(config, null, 2),
      });

      const { status, stderr, stdout } = spawnTyche(fixture, ["--install --target 4.9"], {
        ["TSTYCHE_STORE_PATH"]: "./.store",
      });

      expect(stdout).toMatchInlineSnapshot(`
        "adds TypeScript 4.9.5 to <<cwd>>/tests/__fixtures__/command-line-options/.store/4.9.5
        "
      `);
      expect(stderr).toBe("");

      expect(status).toBe(0);
    });

    test("handles 'current' target received from the configuration file", async () => {
      const config = { target: ["current"] };

      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: isStringTestText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
        ["tstyche.config.json"]: JSON.stringify(config, null, 2),
      });

      const { status, stderr, stdout } = spawnTyche(fixture, ["--install"], {
        ["TSTYCHE_STORE_PATH"]: "./.store",
      });

      expect(stdout).toBe("");
      expect(stderr).toBe("");

      expect(status).toBe(0);
    });

    test("handles 'current' target received from the command line", async () => {
      const config = { target: ["5.0", "latest"] };

      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: isStringTestText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
        ["tstyche.config.json"]: JSON.stringify(config, null, 2),
      });

      const { status, stderr, stdout } = spawnTyche(fixture, ["--install --target current"], {
        ["TSTYCHE_STORE_PATH"]: "./.store",
      });

      expect(stdout).toBe("");
      expect(stderr).toBe("");

      expect(status).toBe(0);
    });
  });

  describe("'--only' option", () => {
    test("selects tests to run", async () => {
      const testText = `import { expect, test } from "tstyche";
test("external is string?", () => {
  expect<string>().type.toBeString();
});

test("external is number?", () => {
  expect<number>().type.toBeNumber();
});

test("internal is string?", () => {
  expect<string>().type.toBeString();
});
`;

      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: testText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      });

      const { status, stderr, stdout } = spawnTyche(fixture, ["--only external"]);

      expect(stdout).toMatchSnapshot("stdout");
      expect(stderr).toBe("");

      expect(status).toBe(0);
    });

    test("selects test group to run", async () => {
      const testText = `import { describe, expect, test } from "tstyche";
describe("external", () => {
  test("is string?", () => {
    expect<string>().type.toBeString();
  });
});

test("external is number?", () => {
  expect<number>().type.toBeNumber();
});

test("internal is string?", () => {
  expect<string>().type.toBeString();
});
`;

      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: testText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      });

      const { status, stderr, stdout } = spawnTyche(fixture, ["--only external"]);

      expect(stdout).toMatchSnapshot("stdout");
      expect(stderr).toBe("");

      expect(status).toBe(0);
    });

    test("does not override the '.skip' run mode flag", async () => {
      const testText = `import { expect, test } from "tstyche";
test("external is string?", () => {
  expect<string>().type.toBeString();
});

test.skip("external is number?", () => {
  expect<number>().type.toBeNumber();
});

test("internal is string?", () => {
  expect<string>().type.toBeString();
});
`;

      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: testText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      });

      const { status, stderr, stdout } = spawnTyche(fixture, ["--only external"]);

      expect(stdout).toMatchSnapshot("stdout");
      expect(stderr).toBe("");

      expect(status).toBe(0);
    });

    test("with '--skip' command line options", async () => {
      const testText = `import { expect, test } from "tstyche";
test("external is string?", () => {
  expect<string>().type.toBeString();
});

test("external is number?", () => {
  expect<number>().type.toBeNumber();
});

test("internal is string?", () => {
  expect<string>().type.toBeString();
});
`;

      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: testText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      });

      const { status, stderr, stdout } = spawnTyche(fixture, ["--only external", "--skip number"]);

      expect(stdout).toMatchSnapshot("stdout");
      expect(stderr).toBe("");

      expect(status).toBe(0);
    });
  });

  describe("'--prune' option", () => {
    test("removes store directory", async () => {
      const storeManifest = { $version: "0" };
      const storeUrl = new URL(".store/", getFixtureUrl(fixture));

      await writeFixture(fixture, {
        [".store/store-manifest.json"]: JSON.stringify(storeManifest),
        ["__typetests__/dummy.test.ts"]: isStringTestText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      });

      expect(existsSync(storeUrl)).toBe(true);

      const { status, stderr, stdout } = spawnTyche(fixture, ["--prune"], { ["TSTYCHE_STORE_PATH"]: "./.store" });

      expect(existsSync(storeUrl)).toBe(false);

      expect(stdout).toBe("");
      expect(stderr).toBe("");

      expect(status).toBe(0);
    });

    test("does nothing, if directory does not exist", async () => {
      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: isStringTestText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      });

      const { status, stderr, stdout } = spawnTyche(fixture, ["--prune"], { ["TSTYCHE_STORE_PATH"]: "./.store" });

      expect(stdout).toBe("");
      expect(stderr).toBe("");

      expect(status).toBe(0);
    });
  });

  describe("'--skip' option", () => {
    test("selects tests to run", async () => {
      const testText = `import { expect, test } from "tstyche";
test("internal is string?", () => {
  expect<string>().type.toBeString();
});

test("internal is number?", () => {
  expect<number>().type.toBeNumber();
});

test("external is string?", () => {
  expect<string>().type.toBeString();
});
`;

      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: testText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      });

      const { status, stderr, stdout } = spawnTyche(fixture, ["--skip internal"]);

      expect(stdout).toMatchSnapshot("stdout");
      expect(stderr).toBe("");

      expect(status).toBe(0);
    });

    test("selects test group to run", async () => {
      const testText = `import { describe, expect, test } from "tstyche";
describe("internal", () => {
  test("is string?", () => {
    expect<string>().type.toBeString();
  });
});

test("internal is number?", () => {
  expect<number>().type.toBeNumber();
});

test("external is string?", () => {
  expect<string>().type.toBeString();
});
`;

      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: testText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      });

      const { status, stderr, stdout } = spawnTyche(fixture, ["--skip internal"]);

      expect(stdout).toMatchSnapshot("stdout");
      expect(stderr).toBe("");

      expect(status).toBe(0);
    });

    test("overrides the '.only' run mode flag", async () => {
      const testText = `import { expect, test } from "tstyche";
test.only("internal is string?", () => {
  expect<string>().type.toBeString();
});

test.only("internal is number?", () => {
  expect<number>().type.toBeNumber();
});

test.only("external is string?", () => {
  expect<string>().type.toBeString();
});
`;

      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: testText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      });

      const { status, stderr, stdout } = spawnTyche(fixture, ["--skip internal"]);

      expect(stdout).toMatchSnapshot("stdout");
      expect(stderr).toBe("");

      expect(status).toBe(0);
    });

    test("with '--only' command line options", async () => {
      const testText = `import { expect, test } from "tstyche";
test("internal is string?", () => {
  expect<string>().type.toBeString();
});

test("internal is number?", () => {
  expect<number>().type.toBeNumber();
});

test.only("external is string?", () => {
  expect<string>().type.toBeString();
});
`;

      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: testText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      });

      const { status, stderr, stdout } = spawnTyche(fixture, ["--only number", "--skip internal"]);

      expect(stdout).toMatchSnapshot("stdout");
      expect(stderr).toBe("");

      expect(status).toBe(0);
    });
  });

  describe("'--target' option", () => {
    test("handles single target", async () => {
      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: isStringTestText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      });

      const { status, stderr, stdout } = spawnTyche(fixture, ["--target 4.8"], { ["TSTYCHE_STORE_PATH"]: "./.store" });

      expect(stdout).toMatchSnapshot("stdout");
      expect(stderr).toBe("");

      expect(status).toBe(0);
    });

    test("handles multiple targets", async () => {
      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: isStringTestText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      });

      const { status, stderr, stdout } = spawnTyche(fixture, ["--target 4.8,5.3.2,current"], {
        ["TSTYCHE_STORE_PATH"]: "./.store",
      });

      expect(stdout).toMatchSnapshot("stdout");
      expect(stderr).toBe("");

      expect(status).toBe(0);
    });

    test("handles 'current' tag", async () => {
      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: isStringTestText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      });

      const { status, stderr } = spawnTyche(fixture, ["--target current"]);

      expect(stderr).toBe("");

      expect(status).toBe(0);
    });

    test("handles not supported '--target' option value", async () => {
      await writeFixture(fixture, {
        ["__typetests__/dummy.test.ts"]: isStringTestText,
        ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      });

      const { status, stderr, stdout } = spawnTyche(fixture, ["--target new"]);

      expect(stdout).toBe("");
      expect(stderr).toMatch(
        [
          "Error: TypeScript version 'new' is not supported.",
          "",
          "Argument for the '--target' option must be a single tag or a comma separated list.",
          "Usage examples:",
        ].join("\r\n"),
      );

      expect(status).toBe(1);
    });
  });

  test("'--version' option", async () => {
    await writeFixture(fixture, {
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { status, stderr, stdout } = spawnTyche(fixture, ["--version"]);

    expect(stdout).toMatch(/^\d+\.\d+\.\d+/);
    expect(stderr).toBe("");

    expect(status).toBe(0);
  });
});
