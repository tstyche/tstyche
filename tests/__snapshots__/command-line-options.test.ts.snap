// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command line options '--target' option: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/dummy.test.ts
  + is string?

uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/dummy.test.ts
  + is string?

Targets:    2 passed, 2 total
Test files: 2 passed, 2 total
Tests:      2 passed, 2 total
Assertions: 2 passed, 2 total
Duration:   <<timestamp>>

Ran all test files.
"
`;
